envi/archs/rx/disasm.py

```diff
@@ -0,0 +1,1349 @@
+            ret.append((self.va + self.size, flags|envi.BR_FALL))
+
+        # if we have no operands, it has no further branches...
+        if len(self.opers) == 0:
+            return ret
+
+        # Check for a call...
+        if self.iflags & IF_CALL:
+            flags |= envi.BR_PROC
+            addb = True
+
+        if addb:
+            oper0 = self.opers[0]
+            if oper0.isDeref():
+                flags |= envi.BR_DEREF
+                tova = oper0.getOperAddr(self, emu=emu)
+            else:
+                tova = oper0.getOperValue(self, emu=emu)
+
+            ret.append((tova, flags))
+
+        return ret
+
+#FIXME: This should just be wrapped into envi
+def addrToName(mcanv, va):
+    sym = mcanv.syms.getSymByAddr(va)
+    if sym is not None:
+        return repr(sym)
+    return "0x%.4x" % va
+
+def renderPossibleLocation(mcanv, op, idx, value):
+    hint = mcanv.syms.getSymHint(op.va, idx)
+    if hint is not None:
+        if mcanv.mem.isValidPointer(value):
+            mcanv.addVaText(hint, value)
+        else:
+            mcanv.addNameText(hint)
+    elif mcanv.mem.isValidPointer(value):
+        name = addrToName(mcanv, value)
+        mcanv.addVaText(name, value)
+    else:
+        value &= 0xffff
+        mcanv.addNameText('0x%x' % value)
+
+class RxRegDirectOper(envi.RegisterOper):
+    '''
+    register operand
+    '''
+
+    def __init__(self, reg, va=0, oflags=0):
+        if reg is None:
+            raise envi.InvalidInstruction(mesg="None Reg Type!",
+                    bytez='f00!', va=va)
+        self.va = va
+        self.reg = reg
+        self.oflags = oflags
+
+    def __eq__(self, oper):
+        if not isinstance(oper, self.__class__):
+            return False
+        if self.reg != oper.reg:
+            return False
+        if self.oflags != oper.oflags:
+            return False
+        return True
+    
+    def involvesPC(self):
+        return self.reg == regs.REG_PC
+
+    def getWidth(self):
+        return regs.rctx.getRegisterWidth(self.reg) // 8
+
+    def getOperValue(self, op, emu=None, codeflow=False):
+        if self.reg == regs.REG_PC and not codeflow:
+            return self.va
+
+        if emu is None:
+            return None
+        return emu.getRegister(self.reg)
+
+    def setOperValue(self, op, emu=None, val=None):
+        if emu is None:
+            return None
+        emu.setRegister(self.reg, val)
+
+    def __eq__(self, oper):
+        if not isinstance(oper, self.__class__):
+            return False
+        if not self.reg == oper.reg:
+            return False
+        if not self.oflags == oper.oflags:
+            return False
+        return True
+
+
+
+class RxDisasm:
+    def __init__(self):
+        self._dis_regctx = regs.RxRegisterContext()
+        self._dis_oparch = envi.ARCH_RX
+
+        self.HANDLERS = [None for x in range(len(formconsts))]
+        self.HANDLERS[FORM_AD] = self.form_AD
```

Comment:
I think the values in HANDLERS matches the values in formconsts, so it might be quicker to instantiate this as one bigger list with all the `self.form_*` values in their proper order.?

Reply:
No very. If you think it's easier to maintain and we don't get much of a performance benefit from it, we can leave it as it.